### ISSUE TITLE
Protocol with callable attribute that has a recursive return type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6127]
### Discussed in https://github.com/microsoft/pyright/discussions/6126

<div type='discussions-op-text'>

<sup>Originally posted by **tjsmart** October 11, 2023</sup>
I have a use case for creating a protocol which has an attribute (`func` in my example below) that will return another instance of this same protocol. Below is a script where I have two attempts at defining such a protocol `CallProto` and `MethProto` along with two possible implementations that I would like to pass the type check `Impl1` and `Impl2`.

I'm running into some errors that I don't think I quite understand and could use some help. Specifically,
1. what do these pyright errors mean?
2. how can I write a protocol that accepts `Impl1` and `Impl2`?
3. when should I use Callable vs. defining a method in a protocol definition?

Links to documentation are welcome 🙂.

Here is my script (pyright errors are given below and inlayed as comments):
```python
from __future__ import annotations
from collections.abc import Callable
from typing import Protocol

class CallProto(Protocol):
    func: Callable[[], CallProto]

class MethProto(Protocol):
    def func(self) -> MethProto:
        ...

class Impl1:
    def func(self) -> Impl1:
        return Impl1()

class Impl2:
    class CallableClass:
        def __call__(self) -> Impl2:
            return Impl2()

    func = CallableClass()


def requires_call_proto(a: CallProto) -> None:
    ...

def requires_meth_proto(b: MethProto) -> None:
    ...

requires_call_proto(Impl1())  # Argument of type "Impl1" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
requires_meth_proto(Impl1())

requires_call_proto(Impl2())  # Argument of type "Impl2" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
requires_meth_proto(Impl2())  # Argument of type "Impl2" cannot be assigned to parameter "b" of type "MethProto" in function "requires_meth_proto"
```

Here is the full pyright output:
```console
$ pyright t.py
/tmp/pyright/t.py
  /tmp/pyright/t.py:30:21 - error: Argument of type "Impl1" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
    "Impl1" is incompatible with protocol "CallProto"
      "func" is invariant because it is mutable
      "func" is an incompatible type
        Type "() -> Impl1" cannot be assigned to type "() -> CallProto"
          Function return type "Impl1" is incompatible with type "CallProto"
            "Impl1" is incompatible with protocol "CallProto" (reportGeneralTypeIssues)
  /tmp/pyright/t.py:33:21 - error: Argument of type "Impl2" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
    "Impl2" is incompatible with protocol "CallProto"
      "func" is invariant because it is mutable
      "func" is an incompatible type
        Type "CallableClass" cannot be assigned to type "() -> CallProto"
          Function return type "Impl2" is incompatible with type "CallProto"
            "Impl2" is incompatible with protocol "CallProto" (reportGeneralTypeIssues)
  /tmp/pyright/t.py:34:21 - error: Argument of type "Impl2" cannot be assigned to parameter "b" of type "MethProto" in function "requires_meth_proto"
    "Impl2" is incompatible with protocol "MethProto"
      "func" is an incompatible type
        Type "CallableClass" cannot be assigned to type "(self: Impl2) -> MethProto"
          Function accepts too many positional parameters; expected 0 but received 1 (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations
```

Other details:
```console
$ lsb_release --description
Description:    Ubuntu 20.04.6 LTS

$ python --version
Python 3.10.13

$ pyright --version
pyright 1.1.316
```

While I prefer to use `pyright` I also checked the output of `mypy` out of curiousity:
```console
$ mypy t.py
t.py:30: error: Argument 1 to "requires_call_proto" has incompatible type "Impl1"; expected "CallProto"  [arg-type]
t.py:30: note: Following member(s) of "Impl1" have conflicts:
t.py:30: note:     Expected:
t.py:30: note:         def () -> CallProto
t.py:30: note:     Got:
t.py:30: note:         def func(self) -> Impl1
t.py:30: note: Protocol member CallProto.func expected settable variable, got read-only attribute
t.py:33: error: Argument 1 to "requires_call_proto" has incompatible type "Impl2"; expected "CallProto"  [arg-type]
t.py:33: note: Following member(s) of "Impl2" have conflicts:
t.py:33: note:     func: expected "Callable[[], CallProto]", got "CallableClass"
Found 2 errors in 1 file (checked 1 source file)
```

In this case `mypy` accepts both `Impl1` and `Impl2` as implementations of `MethProto` which is what I'm hoping for but I don't know if that is correct. 🤷‍♂️</div>

**Comment by erictraut** [comment]
This is included in pyright 1.1.332, which I just published. It will also be included in a future release of pylance.
